File: platform/macos/Bridge/NativeEncoding.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation
import TidyVNC

extension tidyvnc_encoding_schema: ABIValue {}
extension tidyvnc_encoding_choice: ABIValue {}
extension tidyvnc_encoding_value: ABIValue {}
public enum NativeEncodingOption: UInt32, CaseIterable, Sendable {
  case autoSelect = 0, fullColor, lowColorLevel, preferred, customCompression, compression, noJPEG, quality
}
public enum NativeOptionSource: UInt32, Sendable {
  case compiled = 0, appDefaults, profile, session, commandLine, document
}
public struct NativeEncodingAssignment: Sendable {
  public let name: String, value: String
  public init(_ name: String, _ value: String) { self.name = name; self.value = value }
}
public struct NativeEncodingSchema: Sendable {
  public enum Kind: UInt32, Sendable { case boolean = 0, integer, enumeration }
  public let id: NativeEncodingOption, kind: Kind
  public let name: String, alias: String, defaultValue: String
  public let minimum: Int32, maximum: Int32
  public let persistent: Bool, live: Bool
}
public struct NativeEncodingChoice: Sendable {
  public let name: String
  public let wireEncoding: Int32
  public let available: Bool
}
public struct NativeEncodingValue: Equatable, Sendable {
  public let value: String
  public let source: NativeOptionSource
}
private func encodingText<T>(_ value: T) -> String {
  withUnsafeBytes(of: value) { String(decoding: $0.prefix(while: { $0 != 0 }), as: UTF8.self) }
}
// The C handle owns an immutable fixed-size core snapshot. Registry references
// are synchronized; no borrowed span or mutable state escapes into Swift.
public final class NativeEncodingOptions: @unchecked Sendable {
  let handle: NativeHandle
  init(owning raw: UInt64) { handle = NativeHandle(adopting: raw) }
  public convenience init(patch: [NativeEncodingAssignment] = [], source: NativeOptionSource = .session) throws {
    try self.init(base: nil, patch: patch, source: source)
  }
  private init(base: NativeEncodingOptions?, patch: [NativeEncodingAssignment], source: NativeOptionSource) throws {
    guard patch.count <= 256 else { throw NativeError(.resourceLimit, "Too many encoding assignments") }
    var storage: [UInt8] = []
    var offsets: [(Int, Int, Int, Int)] = []
    for item in patch {
      guard item.name.utf8.prefix(129).count <= 128, item.value.utf8.prefix(129).count <= 128 else {
        throw NativeError(.resourceLimit, "Encoding assignment exceeds its byte limit")
      }
      let name = storage.count; storage.append(contentsOf: item.name.utf8)
      let value = storage.count; storage.append(contentsOf: item.value.utf8)
      offsets.append((name, value - name, value, storage.count - value))
    }
    var raw: UInt64 = 0
    try storage.withUnsafeBufferPointer { buffer in
      let assignments = offsets.map { name, nameLength, value, valueLength in
        tidyvnc_encoding_assignment(name: tidyvnc_bytes(data: buffer.baseAddress?.advanced(by: name), length: UInt64(nameLength)),
          value: tidyvnc_bytes(data: buffer.baseAddress?.advanced(by: value), length: UInt64(valueLength)))
      }
      _ = try assignments.withUnsafeBufferPointer { input in
        try checked { tidyvnc_encoding_create(base?.handle.raw ?? 0, input.baseAddress, UInt32(input.count), source.rawValue, &raw, $0) }
      }
    }
    handle = NativeHandle(adopting: raw)
  }
  public func applying(_ patch: [NativeEncodingAssignment], source: NativeOptionSource) throws -> NativeEncodingOptions {
    try NativeEncodingOptions(base: self, patch: patch, source: source)
  }
  public func value(for option: NativeEncodingOption) throws -> NativeEncodingValue {
    var result = abi(tidyvnc_encoding_value.self)
    try checked { tidyvnc_encoding_get(handle.raw, option.rawValue, &result, $0) }
    guard let source = NativeOptionSource(rawValue: result.source) else { throw NativeError(.unsupported, "Unknown option source") }
    return NativeEncodingValue(value: encodingText(result.value), source: source)
  }
  public static func schema() throws -> [NativeEncodingSchema] {
    var result: [NativeEncodingSchema] = []
    for index in 0..<256 {
      var value = abi(tidyvnc_encoding_schema.self)
      if try checked(allowing: [.ok, .noChange], { tidyvnc_encoding_schema_at(UInt32(index), &value, $0) }) == .noChange { return result }
      guard let id = NativeEncodingOption(rawValue: value.id), let kind = NativeEncodingSchema.Kind(rawValue: value.type) else {
        throw NativeError(.unsupported, "Unknown encoding schema entry")
      }
      result.append(NativeEncodingSchema(id: id, kind: kind, name: encodingText(value.name), alias: encodingText(value.alias),
        defaultValue: encodingText(value.default_value), minimum: value.minimum, maximum: value.maximum,
        persistent: value.persistent != 0, live: value.live != 0))
    }
    throw NativeError(.resourceLimit, "Encoding schema exceeds its entry limit")
  }
  public static func choices() throws -> [NativeEncodingChoice] {
    var result: [NativeEncodingChoice] = []
    for index in 0..<256 {
      var value = abi(tidyvnc_encoding_choice.self)
      if try checked(allowing: [.ok, .noChange], { tidyvnc_encoding_choice_at(UInt32(index), &value, $0) }) == .noChange { return result }
      result.append(NativeEncodingChoice(name: encodingText(value.name), wireEncoding: value.wire_encoding, available: value.available != 0))
    }
    throw NativeError(.resourceLimit, "Encoding choices exceed their entry limit")
  }
}
public enum NativeEncodingProblem: UInt32, Sendable {
  case unknownOption = 1, invalidValue, unavailable, tooLong
}
extension NativeError {
  public var encodingProblem: NativeEncodingProblem? {
    domain == UInt32(TIDYVNC_DOMAIN_ENCODING) ? NativeEncodingProblem(rawValue: detail & 0xffff) : nil
  }
  public var encodingOption: NativeEncodingOption? {
    guard domain == UInt32(TIDYVNC_DOMAIN_ENCODING), detail >> 16 > 0 else { return nil }
    return NativeEncodingOption(rawValue: (detail >> 16) - 1)
  }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeEncoding.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

public enum NativeEncodingOption : uint
{
    AutoSelect = 0, FullColor, LowColorLevel, Preferred, CustomCompression, Compression, NoJpeg, Quality,
}

public readonly record struct NativeEncodingAssignment(string Name, string Value);

public sealed record NativeEncodingSchema(NativeEncodingOption Id, NativeEncodingSchema.SchemaKind Kind, string Name, string Alias,
                                          string DefaultValue, int Minimum, int Maximum, bool Persistent, bool Live)
{
    public enum SchemaKind : uint { Boolean = 0, Integer, Enumeration }
}

public sealed record NativeEncodingChoice(string Name, int WireEncoding, bool Available);

public readonly record struct NativeEncodingValue(string Value, NativeOptionSource Source);

public enum NativeEncodingProblem : uint { UnknownOption = 1, InvalidValue, Unavailable, TooLong }

/// <summary>An immutable encoding option snapshot owned by the core (NativeEncodingOptions).</summary>
public sealed class NativeEncodingOptions : IDisposable
{
    internal NativeHandle Handle { get; }

    internal NativeEncodingOptions(NativeHandle handle) => Handle = handle;

    public NativeEncodingOptions(IReadOnlyList<NativeEncodingAssignment>? patch = null,
                                 NativeOptionSource source = NativeOptionSource.Session)
        : this(Create(null, patch ?? [], source)) { }

    private static unsafe NativeHandle Create(NativeEncodingOptions? baseline, IReadOnlyList<NativeEncodingAssignment> patch,
                                              NativeOptionSource source)
    {
        if (patch.Count > 256) throw new NativeError(NativeStatus.ResourceLimit, "Too many encoding assignments");
        var names = patch.Select(item => AbiText.Utf8(item.Name)).ToArray();
        var values = patch.Select(item => AbiText.Utf8(item.Value)).ToArray();
        if (names.Any(n => n.Length > 128) || values.Any(v => v.Length > 128))
            throw new NativeError(NativeStatus.ResourceLimit, "Encoding assignment exceeds its byte limit");
        var handles = new System.Runtime.InteropServices.GCHandle[names.Length * 2];
        try
        {
            var assignments = new tidyvnc_encoding_assignment[names.Length];
            for (var i = 0; i < names.Length; i++)
            {
                handles[2 * i] = System.Runtime.InteropServices.GCHandle.Alloc(names[i], System.Runtime.InteropServices.GCHandleType.Pinned);
                handles[2 * i + 1] = System.Runtime.InteropServices.GCHandle.Alloc(values[i], System.Runtime.InteropServices.GCHandleType.Pinned);
                assignments[i].name = AbiText.Span((byte*)handles[2 * i].AddrOfPinnedObject(), names[i].Length);
                assignments[i].value = AbiText.Span((byte*)handles[2 * i + 1].AddrOfPinnedObject(), values[i].Length);
            }
            var error = Abi.Init<tidyvnc_error>();
            ulong raw = 0;
            fixed (tidyvnc_encoding_assignment* input = assignments)
                Abi.Check(NativeMethods.tidyvnc_encoding_create(baseline?.Handle.Raw ?? 0, input, (uint)assignments.Length,
                    (uint)source, &raw, &error), &error);
            return NativeHandle.Adopt(raw);
        }
        finally
        {
            foreach (var pinned in handles) if (pinned.IsAllocated) pinned.Free();
        }
    }

    public NativeEncodingOptions Applying(IReadOnlyList<NativeEncodingAssignment> patch, NativeOptionSource source)
        => new(Create(this, patch, source));

    public unsafe NativeEncodingValue Value(NativeEncodingOption option)
    {
        var result = Abi.Init<tidyvnc_encoding_value>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_encoding_get(Handle.Raw, (uint)option, &result, &error), &error);
        if (!Enum.IsDefined((NativeOptionSource)result.source)) throw new NativeError(NativeStatus.Unsupported, "Unknown option source");
        return new NativeEncodingValue(AbiText.Fixed(result.value, 32), (NativeOptionSource)result.source);
    }

    public static unsafe IReadOnlyList<NativeEncodingSchema> Schema()
    {
        var result = new List<NativeEncodingSchema>();
        var error = Abi.Init<tidyvnc_error>();
        for (uint index = 0; index < 256; index++)
        {
            var value = Abi.Init<tidyvnc_encoding_schema>();
            if (Abi.Check(NativeMethods.tidyvnc_encoding_schema_at(index, &value, &error), &error, NativeStatus.Ok, NativeStatus.NoChange) == NativeStatus.NoChange)
                return result;
            if (!Enum.IsDefined((NativeEncodingOption)value.id) || !Enum.IsDefined((NativeEncodingSchema.SchemaKind)value.type))
                throw new NativeError(NativeStatus.Unsupported, "Unknown encoding schema entry");
            result.Add(new NativeEncodingSchema((NativeEncodingOption)value.id, (NativeEncodingSchema.SchemaKind)value.type,
                AbiText.Fixed(value.name, 32), AbiText.Fixed(value.alias, 32), AbiText.Fixed(value.default_value, 32),
                value.minimum, value.maximum, value.persistent != 0, value.live != 0));
        }
        throw new NativeError(NativeStatus.ResourceLimit, "Encoding schema exceeds its entry limit");
    }

    public static unsafe IReadOnlyList<NativeEncodingChoice> Choices()
    {
        var result = new List<NativeEncodingChoice>();
        var error = Abi.Init<tidyvnc_error>();
        for (uint index = 0; index < 256; index++)
        {
            var value = Abi.Init<tidyvnc_encoding_choice>();
            if (Abi.Check(NativeMethods.tidyvnc_encoding_choice_at(index, &value, &error), &error, NativeStatus.Ok, NativeStatus.NoChange) == NativeStatus.NoChange)
                return result;
            result.Add(new NativeEncodingChoice(AbiText.Fixed(value.name, 32), value.wire_encoding, value.available != 0));
        }
        throw new NativeError(NativeStatus.ResourceLimit, "Encoding choices exceed their entry limit");
    }

    public void Dispose() => Handle.Dispose();
}

public static class NativeEncodingErrors
{
    public static NativeEncodingProblem? EncodingProblem(this NativeError error)
        => error.Domain == Tidyvnc.TIDYVNC_DOMAIN_ENCODING && Enum.IsDefined((NativeEncodingProblem)(error.Detail & 0xffff))
            ? (NativeEncodingProblem)(error.Detail & 0xffff) : null;

    public static NativeEncodingOption? EncodingOption(this NativeError error)
        => error.Domain == Tidyvnc.TIDYVNC_DOMAIN_ENCODING && error.Detail >> 16 > 0
            ? (NativeEncodingOption)((error.Detail >> 16) - 1) : null;
}
